[AutoDiff] Don't try to differentiate generic functions before it's supported

rxwei opened the pull request:
Differentiating generic functions is not supported yet. This patch teaches AD to bail out instead of running primal synthesis on the generic function, which might lead to either zero derivatives or a crasher.

Zero derivatives are caused by incorrect activity analysis. Activity analysis does not recognize loads, stores, and indirect returns yet.

This patch does not cover all cases (e.g. zero derivatives) because we really need to fix activity analysis first. I'll send a more comprehensive patch that will cover more test cases later.

rxwei committed:
Diagnose differentiating generic functions

rxwei commented:
@swift-ci please test tensorflow

dan-zheng commented on lib/SILOptimizer/Mandatory/TFDifferentiation.cpp:
```diff
@@ -2407,11 +2406,24 @@ class PrimalGenCloner final : public SILClonerWithScopes<PrimalGenCloner> {
+    auto *newTask =
+        context.lookUpMinimalDifferentiationTask(calleeOriginalFn, indices);
+    if (!newTask) {
+      // If the callee function does not have a primitive declaration that
```
Nit: I'd switch clause order.
```suggestion
      // If the callee function is generic and does not have a primitive
      // declaration that covers the required indices, we cannot differentiate it
      // yet so we bail out instead of registering a new differentiation task.
```

dan-zheng reviewed (approved):
Why not just fix generics AD? 😛
Totally kidding, LGTM.

dan-zheng committed:
Update lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

Co-Authored-By: rxwei <rxwei@google.com>

rxwei committed:
Switch to a simpler fix.

Instead of preventing the generic function from being registered, we just reject it in `PrimalGen::performSynthesis`, where we also detect and reject control flow functions.

Also fix the "unchecked or non-canonical type" assertion in primal value struct creation. The struct should share the same generic environment with the original function.

rxwei committed:
Merge branch 'fix-ad-crasher-2' of github.com:rxwei/swift into fix-ad-crasher-2

rxwei committed:
Revert some unnecessary changes.

rxwei commented:
I switched to a simpler fix: Instead of preventing the generic function from being registered, we just reject it in `PrimalGen::performSynthesis`. This is where we also detect and reject control flow functions.

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow
